### ISSUE TITLE
Untrusted Agents

### ISSUE THREAD

**Issue by Saturma** [mitre/caldera#2252]
Hello. 

When I execute some operations, some abilities are skipped because the agent is untrusted. 

![image](https://user-images.githubusercontent.com/23501646/132491737-f9cd7283-0fbd-472f-9b28-4410d68a5ccc.png)

I just restarted the agent so it must be trusted. The comunications are fine. 

![image](https://user-images.githubusercontent.com/23501646/132492033-0ca4ec2f-4cf0-4d0c-987c-8b42bbc6711a.png)

I can't see anything in the logs.  Is there a way to change the agent's status to trusted? How can I get the reason why the agent is becoming untrusted? 

The agent is deployed in a Windows 10.  
Caldera Server: 3.1.0
Agent: Sandcat

Regards

**Comment by Saturma** [comment]
hi @uruwhy. 

Yes, I tried your solution but with the same result. I don't know why the agent is marked as untrusted.

**Comment by uruwhy** [comment]
When you removed the first untrusted agent from the table, the agent popped back up in the table, and then it became untrusted again?

**Comment by uruwhy** [comment]
Just to verify, you removed the agent from the table, it reappeared as trusted, and then it became untrusted? If you started your C2 server with debug logging, could you share screenshots of that, as well?

**Comment by alexanderkent** [comment]
In addition, under `Agents` in the CALDERA web ui, try changing the `untrusted timer` value to a generous value e.g. instead of 90 (default) try 9000. This would take affect for new agent deployments and not for currently deployed agents.

**Comment by Saturma** [comment]
I have 3 machines: Kali linux (Caldera Server) and 2 Windows 10. 

Caldera server runs with: python3 server.py --insecure 

I install a Sandcat Caldera Agent (Black.exe) in one W10 machine and I want to do a Lateral Movement to the other Windows 10 machine. 
CMD: .\Black.exe -server http://10.10.20.13:8888 -v

The agent is always green.  and with comunications.
![image](https://user-images.githubusercontent.com/23501646/133107838-8eb3dded-608d-48c8-a447-34529bc7c9ee.png)

I run the Adversary "Service Creation Lateral movement"

![image](https://user-images.githubusercontent.com/23501646/133108222-9b96e99e-1767-403d-b685-7fea0eb65041.png)

I create a Fact Source with all the needed variables set. 
![image](https://user-images.githubusercontent.com/23501646/133108699-8b9b4958-6c65-4c24-bb02-0617d6ca59a8.png)

I run the Operation "LM"
![image](https://user-images.githubusercontent.com/23501646/133108924-4a39e613-4293-4e9a-ae04-7e8fd9449c79.png)

I only see two steps: 
![image](https://user-images.githubusercontent.com/23501646/133109309-5fa3072d-fee6-40b9-8f02-7435f4bd2e1b.png)

The copy doesn´t appear. It doesn´t work, 

If I open the report... 
![image](https://user-images.githubusercontent.com/23501646/133109681-84934e9e-1871-4c07-a947-e5b76fdc131d.png)

Sometimes it says: 
![image](https://user-images.githubusercontent.com/23501646/133111487-bf8a71e9-743d-4cf2-b22c-cb9b3c68ba29.png)


I remove the agent with the red cross, and the agent appear again.

![image](https://user-images.githubusercontent.com/23501646/133110203-cacc4c63-827b-4f45-8bb1-90984375dc7a.png)

Where is the untrusted Flag??

**Comment by uruwhy** [comment]
If the agent is trusted, it'll have the green box for the PID in the agent table. If it's untrusted, that box will turn red. If you run your server with debug logging (e.g. `python3 server.py --insecure --log DEBUG`), you'll get debug notifications when agents become untrusted.

As for why the agent is being untrusted, you can try making sure that the process in the powershell window doesn't hang. Sometimes when you run programs directly in the powershell window like that, clicking on it and then clicking elsewhere can pause it occasionally (just press enter to force resuming). I've personally run into that issue a bunch of times - you can also just run it as a background process if you don't want to deal with that, though you'd have to redirect output to a log file and tail it from there.

**Comment by Saturma** [comment]
The issue seems to be in the Fact Sources.  The machine where I want to move to is called "MES"

I have a Fact Source: 
![image](https://user-images.githubusercontent.com/23501646/133750297-e0875b1d-dbb9-45cb-acc4-ad43bab6c58a.png)

This does not work. However, if I make a new adversary without using facts or do it by hand, it works. This WORKS!!! 

![image](https://user-images.githubusercontent.com/23501646/133750799-621e9d8e-a9c8-4267-b25c-c863bcbc84e3.png)


BUT THIS NOT (the original)

![image](https://user-images.githubusercontent.com/23501646/133751195-56fc2df6-3a04-4377-a736-7e10cbce863a.png)
![image](https://user-images.githubusercontent.com/23501646/133751277-a4345d6d-003c-4575-83b1-69939d15d487.png)

The only change is the Fact Sources.

**Comment by uruwhy** [comment]
When you start your operation, are you selecting the correct fact source? And if you export the operation report, does it say that the ability was skipped due to missing facts?

**Comment by Saturma** [comment]
Yes, I choose the correct fact source. I did this test many times.

It never says that it was skipped due to missing facts. Only the errors mencioned above: 
![image](https://user-images.githubusercontent.com/23501646/133758600-7d4f5ee6-e76d-4f53-8c5a-e8aad2771be7.png)
![image](https://user-images.githubusercontent.com/23501646/133758626-8aa4008c-2cdb-4b66-b341-ec739fefd499.png)

That's why I was confused about the untrusted agent.

**Comment by uruwhy** [comment]
Couple follow-up questions - next time you restart the server, can you add `--log DEBUG` to the server start command to enable debug-verbosity logging? That way you can see exactly when the agent becomes untrusted. The debug line may look something like:
```
2021-09-17 05:47:04 - DEBUG (app_svc.py:47 start_sniffer_untrusted_agents) Agent (cxzgva) now untrusted. Last seen 106 sec ago
```
As for why the operation suddenly works if you hardcode the machine name into the ability command - could you show the output of the "View Remote Shares" ability? The "Copy Sandcat SMB" ability has many requirements for the `remote.host.fqdn` fact, so even if you have that fact in the fact source, it doesn't guarantee that the fact will be used.

**Comment by radioheado** [comment]
I encountered the same issue. My Caldera server runs on Ubuntu 20.04.2 LTS, and the red agent runs on Windows 10.
I created the red agent through a Java program that I wrote myself on that Windows machine. The agent had a green box with its PID initially, then it suddenly turned red with no reason. Then I tried to execute an operation using REST API, the output shows all abilities are skipped with `"reason_id": 5 `, which is `"Agent untrusted"`. I checked the blue agent that I deployed in the same way, its PID is still in the green box. This is confusing.
I tried to stop the operation from terminal, but failed to find the `operation_id`. This is also a question I want to ask, but there's already an issue posted yesterday #2259, and I'm also waiting for updates. Luckily I can see this operation through web interface. I clicked stop there, waited the automatically generated report and got the `operation_id`. Then I tried to manipulate the same operation from terminal with the correct  `operation_id`, the same error prompted no matter what status command I wanted to execute.
![image](https://user-images.githubusercontent.com/42974719/133824537-cf4b3b1d-c1bb-4124-9e35-60c35a2b40b6.png)

**Comment by uruwhy** [comment]
@radioheado when you say that you created your agent through a Java program, do you mean that you ran the agent executable via a Java program? Or that you wrote a custom Java agent to communicate with the C2 server?

Can you verify that your agent was still actively beaconing to your C2 server (e.g. via tcpdump or wireshark, or server and/or agent logs) when it became untrusted and even afterwards?

Regarding your operation_id question, please see the resolution for https://github.com/mitre/caldera/issues/2259. If that doesn't address your situation, could you open a separate issue for it?
